fxa-rust-client/src/oauth.rs

```diff
@@ -0,0 +1,333 @@
+use errors::*;
+use regex::Regex;
+use url::Url;
+
+// These character ranges are from the OAuth RFC,
+// https://tools.ietf.org/html/rfc6749#section-3.3
+const VALID_SCOPE_VALUE: &str = r"^[\x21\x23-\x5B\x5D-\x7E]+$";
+const VALID_SHORT_NAME_VALUE: &str = r"^[a-zA-Z0-9_]+$";
+const VALID_FRAGMENT_VALUE: &str = r"^[a-zA-Z0-9_]+$";
+
+enum ScopeValue {
+    ShortScope(ShortScopeValue),
+    URLScope(URLScopeValue),
+}
+
+struct ShortScopeValue {
+    has_write: bool,
+    names: Vec<String>,
+}
+
+impl ShortScopeValue {
+    pub fn new(value: &str) -> Result<ShortScopeValue> {
+        if !Regex::new(VALID_SCOPE_VALUE).unwrap().is_match(value) {
+            bail!("Invalid scope value: {}", value);
+        }
+        let mut names = value
+            .split(":")
+            .map(|value| value.to_string())
+            .collect::<Vec<String>>();
+        for name in &names {
+            if !Regex::new(VALID_SHORT_NAME_VALUE).unwrap().is_match(name) {
+                bail!("Invalid scope value: {}", value);
+            }
+        }
+        let has_write = match names.last() {
+            Some(last_item) => match last_item.as_ref() {
+                "write" => {
+                    if names.len() == 0 {
+                        bail!("Invalid scope value: {}", value);
+                    }
+                    true
+                }
+                _ => false,
+            },
+            None => bail!("Empty names!"),
+        };
+        if has_write {
+            names.pop();
+        }
+        Ok(ShortScopeValue { has_write, names })
+    }
+
+    fn implies(&self, value: &ShortScopeValue) -> bool {
+        if value.has_write && !self.has_write {
+            return false;
+        }
+        if value.names.len() < self.names.len() {
+            return false;
+        }
+        self.names
+            .iter()
+            .zip(value.names.iter())
+            .all(|(a, b)| a == b)
+    }
+
+    fn to_string(&self) -> String {
+        let mut str = self.names.join(":");
+        if self.has_write {
+            str.push_str(":write");
+        }
+        str
+    }
+}
+
+struct URLScopeValue {
+    url: Url,
+}
+
+impl URLScopeValue {
+    pub fn new(value: &str) -> Result<URLScopeValue> {
+        if !Regex::new(VALID_SCOPE_VALUE).unwrap().is_match(value) {
+            bail!("Invalid scope value: {}", value);
+        }
+        let url = Url::parse(value)?;
+        if url.scheme() != "https" {
+            bail!("Invalid scope value: {}", value);
+        }
+        if url.username().len() > 0 && url.password().is_some() && url.query().is_some() {
```

Comment:
Should these be `&&` or `||`?  IIUC we want to bail out if any of those conditions are true.